### PR TITLE
[KernelGen][Nvidia] Add _histogramdd_from_bin_cts operator with Triton kernel

### PR DESCRIPTION
**Pull request by Yukun-Cui**
> Replaces #5808 — original could not be reopened after the head branch was rebuilt to drop ~3,400 spurious mirror-history commits (now a clean single commit on ). Content is byte-identical to the original branch tip.

## Summary

Adds a Triton kernel for `_histogramdd_from_bin_cts`. Computes a multi-dimensional histogram of the values in a tensor using a sequence of per-dimension bin counts (equal-width bins).

## Testing
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_histogramdd_from_bin_cts.py --level core` (NVIDIA H20)

### histogramdd_from_bin_cts

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup |
|-------|------|--------------------|-------------------|---------|
| float32 | [1024, 2] | 0.183920 | 0.180720 | 1.018 |
| float32 | [10000, 2] | 0.171536 | 0.196448 | 0.873 |
| float32 | [100000, 2] | 0.292560 | 0.203856 | 1.435 |
| float32 | [1000, 3] | 0.233072 | 0.210304 | 1.108 |
| float32 | [50000, 3] | 0.338464 | 0.225904 | 1.498 |
| float32 | [100000, 3] | 0.433600 | 0.211712 | 2.048 |
| float32 | [10000, 1] | 0.089696 | 0.179680 | 0.499 |
| float32 | [1000000, 1] | 0.982560 | 0.492176 | 1.996 |
| float32 | [1024, 2] | 0.182032 | 0.195136 | 0.933 |
| float32 | [10000, 2] | 0.200928 | 0.194512 | 1.033 |
| float32 | [100000, 2] | 0.358608 | 0.203184 | 1.765 |
| float32 | [1000, 3] | 0.244416 | 0.208576 | 1.172 |
| float32 | [50000, 3] | 0.392992 | 0.207776 | 1.891 |
| float32 | [100000, 3] | 0.553136 | 0.207808 | 2.662 |
| float32 | [10000, 1] | 0.138944 | 0.179856 | 0.773 |
| float32 | [1000000, 1] | 6.018336 | 0.489264 | 12.301 |
| float32 | [1024, 2] | 0.179056 | 0.290304 | 0.617 |
| float32 | [10000, 2] | 0.230432 | 0.291168 | 0.791 |
| float32 | [100000, 2] | 0.311456 | 0.308144 | 1.011 |
| float32 | [1000, 3] | 0.247328 | 0.299376 | 0.826 |
| float32 | [50000, 3] | 0.369408 | 0.304544 | 1.213 |
| float32 | [100000, 3] | 0.460768 | 0.305488 | 1.508 |
| float32 | [10000, 1] | 0.100560 | 0.275408 | 0.365 |
| float32 | [1000000, 1] | 0.988608 | 0.598544 | 1.652 |
| float32 | [1024, 2] | 0.170528 | 0.196672 | 0.867 |
| float32 | [10000, 2] | 0.185616 | 0.195936 | 0.947 |
| float32 | [100000, 2] | 0.299488 | 0.203584 | 1.471 |
| float32 | [1000, 3] | 0.226768 | 0.208832 | 1.086 |
| float32 | [50000, 3] | 0.344544 | 0.208944 | 1.649 |
| float32 | [100000, 3] | 0.447920 | 0.209920 | 2.134 |
| float32 | [10000, 1] | 0.121728 | 0.181408 | 0.671 |
| float32 | [1000000, 1] | 0.992800 | 0.492192 | 2.017 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| histogramdd_from_bin_cts | **1.620** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (H20) | PASS (32 cases) | 1.620 | Primary |
| Tianshu | N/A | — | — |
| Muxi | N/A | — | — |
| Ascend | N/A | — | — |
| Hygon | N/A | — | — |

## Files Changed
- `src/flag_gems/ops/_histogramdd_from_bin_cts.py`: Triton kernel implementation
- `tests/test_histogramdd_from_bin_cts.py`: Accuracy test
- `benchmark/test_histogramdd_from_bin_cts.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Reduction, stage: alpha 5.0)

### CODE REVIEW

File: conf/operators.yaml [hunk 1]
```diff
@@ -6185,6 +6185,20 @@ ops:
...
+    '
+  for:
+  - _histogramdd_from_bin_cts
+  labels:
+  - aten
+  - KernelGen
+  kind:
+  - Reduction
+  stages:
+  - alpha: '5.0'
```
Comment (Caeruleann) [inline conf/operators.yaml]:
```suggestion
  - alpha: '5.4'
```

File: conf/operators.yaml [hunk 2]
```diff
@@ -6185,6 +6185,20 @@ ops:
   - Math
   stages:
   - alpha: '5.3'
+- id: histogramdd_from_bin_cts
```
Comment (Caeruleann) [inline conf/operators.yaml:6188]: The `_histogramdd_from_bin_cts_out` registration is missing.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+    row_mask = row_start < M
+
+    # Whether this thread has a valid point.
+    valid = row_mask
+    linear_idx = tl.zeros((BLOCK_SIZE,), dtype=tl.int64)
+
+    # Walk over each dimension (unrolled at compile time via N_CONST).
+    for d in range(N_CONST):
+        coord = tl.load(
+            inp_ptr + row_start * N_CONST + d, mask=row_mask, other=float("nan")
```
Comment (Caeruleann) [inline src/flag_gems/ops/_histogramdd_from_bin_cts.py:72]: Please preserve float64 throughout the FP64 path. The coordinates, bin boundaries, inverse widths, and density values are currently converted to float32, which can assign precision-sensitive FP64 values to the wrong bin. Please also add FP64 parity tests with values close to bin boundaries.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        # of the per-dimension bin widths = prod((right - left) / bins). Both the
+        # total count and bin widths are tiny scalars, so move them to the host
+        # (one sync) and compute the scale in Python instead of launching extra
+        # CUDA reductions.
+        bin_volume = 1.0
+        widths_host = widths.tolist()
+        for d in _builtins.range(N):
+            bin_volume *= widths_host[d] / bins_list[d]
+        total_count = hist_acc.sum().item()
+        scale = (1.0 / (total_count * bin_volume)) if total_count != 0.0 else 0.0
```
Comment (Caeruleann) [inline src/flag_gems/ops/_histogramdd_from_bin_cts.py:295]: Please do not force the density scale to zero when total_count == 0. ATen preserves the division result here, producing NaNs for empty/all-out-of-range inputs and infinities for some cancelling-weight cases. Please match this behavior and add parity tests.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+                f"range must have 2 values per dimension ({2 * N} total), got {len(rng)}"
+            )
+        lefts = torch.tensor(rng[0::2], dtype=inp.dtype, device=inp.device)
+        rights = torch.tensor(rng[1::2], dtype=inp.dtype, device=inp.device)
+        return lefts, rights
+
+    # Auto-range per dimension using the flattened (M, N) view.
+    flat = inp.reshape(-1, N)
+    # torch.aminmax supports all floating dtypes (incl. float64); amin/amax do not.
+    mins, maxs = torch.aminmax(flat, dim=0)
```
Comment (Caeruleann) [inline src/flag_gems/ops/_histogramdd_from_bin_cts.py:148]: Please move the automatic-range and density computations to Triton kernels or existing FlagGems operators. The current host path calls PyTorch computation operators such as aminmax, where, and sum, which is not allowed by the FlagGems host-function convention.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+    # Allocate / validate the output buffer.
+    if out is not None:
+        if tuple(out.shape) != out_shape:
+            raise RuntimeError(
+                "_histogramdd_from_bin_cts.out: out shape mismatch: expected "
+                f"{out_shape}, got {tuple(out.shape)}"
+            )
+        out = out.to(out_dtype) if out.dtype != out_dtype else out
+    else:
+        out = torch.empty(out_shape, dtype=out_dtype, device=self.device)
```
Comment (Caeruleann) [inline src/flag_gems/ops/_histogramdd_from_bin_cts.py:199]: Please match the full out= contract. ATen resizes a resizable output tensor and rejects a dtype mismatch, while this implementation rejects shape changes and rebinds out through .to(), leaving the caller-provided buffer untouched. Please add resize, dtype-error, and returned-tensor identity tests.

File: conf/operators.yaml (hunk 1 above)
Comment (Yukun-Cui) [reply conf/operators.yaml]: Fixed! Updated to alpha 5.4.

File: conf/operators.yaml (hunk 2 above)
Comment (Yukun-Cui) [reply conf/operators.yaml:6188]: I'll add the missing _histogramdd_from_bin_cts_out registration.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 3 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py:72]: Critical fix! I'll preserve the original dtype throughout the computation and add FP64 precision tests with values close to bin boundaries.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 4 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py:295]: I'll remove the zero-forcing logic and let the natural NaN/Inf propagation happen, matching ATen's semantics. Will add parity tests.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 5 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py:148]: I'll refactor these computations into Triton kernels or use existing FlagGems operators to avoid PyTorch operations in host code.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 6 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py:199]: I'll implement proper out= handling with resize support and dtype validation, plus comprehensive tests.

File: conf/operators.yaml (hunk 1 above)
Comment (Yukun-Cui) [reply conf/operators.yaml]: 已是 `alpha: '5.4'`，这条建议已应用。

File: conf/operators.yaml (hunk 2 above)
Comment (Yukun-Cui) [reply conf/operators.yaml:6991]: 已补上 `_histogramdd_from_bin_cts.out` 的注册（id `histogramdd_from_bin_cts_out`，`for: _histogramdd_from_bin_cts.out`），结构与仓库里其他 `_out` 条目一致。

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 3 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py:79]: 已修，但这条比"不要窄化到 fp32"要深一层，所以把实测过程写清楚。

坐标、边界、bin 宽、density 现在全程保持输入 dtype，不再窄化。但仅此不够——**ATen 不是用算术求 bin 索引的**：它用 `linspace` 生成边界数组再做查找，没有任何闭式公式能复现。实测三种算术形式都对不上：

```
floor((coord-left) * bins/width)  ->  0.3 落 bin 3，ATen 说 2
floor((coord-left) / bin_width)   ->  0.3 修好了，但 0.6 / 0.7 又错
```

所以 kernel 改成构建同样的边界表并二分查找（`histogramdd_edges_kernel` + 展开的 log₂(bins) 步搜索）。

要位级复现 `linspace`，还需要三处修正，每处都有实测的失配计数：

| 规则 | 错误做法的失配数 |
|---|---|
| 从较近端点插值 | 一路从左累加：113/300（边界 6 变成 `0.6000000000000001`，正好落在 `0.6` 的点掉一个 bin） |
| `step` 舍入到存储 dtype，乘加在 fp64 | 全程 fp32：113/300；`step` 由未窄化的 fp64 端点算：167/200 |
| fp32 用 `tl.math.div_rn` | 普通 `/`：47/200 与 IEEE 除法不一致 |

最后一条值得单独提：Triton 把普通 fp32 `/` 降级成快速倒数，不是正确舍入。这足以让 `step` 偏 1 ulp，并带偏所有内部边界。

结果：边界值与 `aten::_histogramdd_bin_edges` 在 200/200 组配置上完全一致（fp32 与 fp64）；**正好落在边界上的点从 4/60 组提升到 60/60 组**。

新增 `test_histogramdd_from_bin_cts_bin_boundary_precision`，对每个边界及其两侧各 1 ulp 的值做精确计数比较（用 `gems_assert_equal`，不是近似）。

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 4 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py]: 已修，实测确认了三种情形，你描述的两类都对上了：

```
全部越界      density=True -> [nan, nan, nan, nan]
空输入 (M=0)  density=True -> [nan, nan, nan, nan]
权重相互抵消              -> [inf, nan, nan, -inf]
```

`total_count == 0` 时强制置 0 的逻辑已删除，scale kernel 现在直接做除法不加保护，NaN/±Inf 自然传播；density kernel 也保持在累加 dtype 中计算，让非有限的 scale 原样传递。

新增两个测试：`test_histogramdd_from_bin_cts_density_zero_total`（全越界 + 空输入）和 `test_histogramdd_from_bin_cts_density_cancelling_weights`。两个测试都先断言参考实现确实产出 NaN/Inf，再比对，这样万一将来 ATen 行为变了，测试会指出原因而不是静默通过。

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 5 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py]: 已全部搬到 Triton kernel。host 侧不再有任何 PyTorch 计算算子调用：

- `torch.aminmax` / `torch.where` → `histogramdd_col_range_kernel`（逐列 min/max，min==max 的退化维度在 kernel 内扩展成单位区间；用 `tl.minimum`/`tl.maximum` 保持 NaN 传播语义，与 aminmax 一致）
- 逐维 `safe_widths[d].item()` → `histogramdd_bin_geometry_kernel`
- `hist_acc.sum().item()` 与 host 侧 `widths.tolist()` 求 bin volume → `histogramdd_log_volume_kernel` + `histogramdd_total_scale_kernel`

bin volume 在对数域累加、符号单独跟踪，避免高维乘积在做除法之前先溢出/下溢。

现在整个 host 函数里只剩张量分配、`reshape`/`contiguous` 这类视图操作，以及 kernel 启动，没有 `.item()`/`.tolist()`/归约调用。

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 6 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py:464]: 已按原生契约实现，实测的 ATen 行为如下：

```
out 形状不符 (0,) 或 (5,7)  -> 原地 resize 成 (3,3)，返回的就是 out 本身
out dtype 为 fp64 / 输入 fp32 -> RuntimeError: Expected out tensor to have dtype float, but got double
out 非连续但形状正确        -> 接受
```

三点都改了：

- 形状不符改为 `out.resize_(out_shape)`，不再报错；返回的始终是调用方传入的对象
- dtype 不符改为抛 `RuntimeError`，不再用 `.to()` 重新绑定——原来那样会让调用方的缓冲区完全没被写入，这是比报错更糟的静默失败
- 非连续 out 无法 `view(num_bins)`，改为在临时缓冲中累加后 `copy_` 回去，保留调用方的 layout

顺带把验证错误也对齐了 ATen 的类型和措辞（`RuntimeError`/`NotImplementedError`，不再是 `ValueError`）。

新增三个测试：`..._out_resize`（从 `(0,)` 和 `(5,7)` 两种起始形状，断言对象同一性）、`..._out_dtype_mismatch`、`..._out_non_contiguous`。

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py [hunk 7]
```diff
@@ -0,0 +1,645 @@
...
+                f"torch.histogramdd: for a {N}-dimensional histogram range should "
+                f"have {2 * N} elements, but got {len(rng)}"
+            )
+        lefts = torch.tensor(rng[0::2], dtype=torch.float64, device=inp.device)
+        rights = torch.tensor(rng[1::2], dtype=torch.float64, device=inp.device)
+        return lefts, rights
+
+    lefts = torch.empty(N, dtype=torch.float64, device=inp.device)
+    rights = torch.empty(N, dtype=torch.float64, device=inp.device)
+    M = inp.shape[0]
```
Comment (Caeruleann) [inline src/flag_gems/ops/_histogramdd_from_bin_cts.py:415]: Native also validates explicit ranges: range=[1.5, -0.5, ...] raises "min should not exceed max, but got min 1.5 max -0.5 for dimension 0", while this silently returns an all-zero histogram. Please reject left > right per dimension with the same error.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py [hunk 8]
```diff
@@ -0,0 +1,645 @@
...
+        # No data to reduce over; ATen's empty-input range collapses to (0, 1) per
+        # dimension after the degenerate-range expansion of (inf, -inf) is skipped.
+        lefts.fill_(0.0)
+        rights.fill_(1.0)
+        return lefts, rights
+
+    BLOCK_N = triton.next_power_of_2(N)
+    grid = (triton.cdiv(N, BLOCK_N),)
+    with torch_device_fn.device(inp.device):
+        histogramdd_col_range_kernel[grid](
```
Comment (Caeruleann) [inline src/flag_gems/ops/_histogramdd_from_bin_cts.py:426]: When the input contains NaN/inf and range=None, the auto-range becomes non-finite: ATen raises "torch.histogramdd: dimension 0's range [nan, nan] is not finite", but this silently returns a garbage histogram. Please add the same finiteness check after the range is resolved and raise before computing.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 7 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py:415]: 已修（04ca47ab）。`range` 现在逐维校验 `left <= right`，在构造 bin edges 之前：

```
torch.histogramdd: min should not exceed max, but got min 1.5 max -0.5 for dimension 0
```

bound 用 ATen 的 `%g` 格式（6 位有效数字）打印——实测确认过，`0.123456789` 输出 `0.123457`、`1e7` 输出 `1e+07`，所以不是 Python 默认 repr。相等（min == max）是允许的，与 native 一致。

新增 `test_histogramdd_from_bin_cts_rejects_inverted_range`（2-D 与 3-D）与一个 `%g` 格式对齐测试，都与 native 消息逐字比对。

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 8 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py:426]: 已修（04ca47ab）。`range=None` 时在设备归约之后加了有限性检查：

```
torch.histogramdd: dimension 0's range [nan, nan] is not finite
```

但这条改动暴露出 range kernel 里一个更深的 bug，值得说明，因为它意味着**修复前即使不报错，NaN 列的 range 也是错的**：

kernel 原来依赖 `tl.min` / `tl.max` / `tl.minimum` / `tl.maximum` 传播 NaN（注释也是这么写的），但它们全都 lower 成 `fmin`/`fmax`，**会忽略 NaN 操作数**。所以：

- NaN 列被折叠成 `+inf/-inf` 恒等元，range 报成 `[-inf, inf]` 而不是 `[nan, nan]`（这正是你看到的 `inf/-inf` 消息的成因）；
- 更隐蔽的是，NaN 落在较早的 row tile 里时会被后续归约直接丢掉（`BLOCK_M=128`，所以 row 0 和 row 200 的表现不同）。

现在 NaN 状态用一个独立的布尔累加器跨 tile 做 OR，最后统一施加到两个 bound 上，与 `aminmax` 语义一致。

验证：NaN/±Inf × 三个 row 位置（含跨 128-row tile 边界）共 18/18 与 native 消息逐字一致。新增 `test_histogramdd_from_bin_cts_auto_range_must_be_finite`（3 个坏值 × 3 个 row 位置）与一个反向测试确认正常有限数据不受影响。

`tests/test_histogramdd_from_bin_cts.py`: 90 passed（`--ref=cpu --quick` 70 passed）。

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py [hunk 9]
```diff
@@ -0,0 +1,702 @@
...
+    """
+    finite = torch.isfinite(lefts) & torch.isfinite(rights)
+    if bool(finite.all()):
+        return
+    dim = int((~finite).nonzero()[0].item())
+    raise RuntimeError(
+        f"torch.histogramdd: dimension {dim}'s range "
+        f"[{_format_bound(lefts[dim].item())}, {_format_bound(rights[dim].item())}] "
+        "is not finite"
+    )
```
Comment (Caeruleann) [inline src/flag_gems/ops/_histogramdd_from_bin_cts.py:458]: This check runs torch.isfinite / nonzero on the host, which brings back the PyTorch compute calls this PR otherwise removed per the host conventions. Please have the col_range kernel write out a finite flag and only read that single scalar here.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py [hunk 10]
```diff
@@ -0,0 +1,702 @@
...
+            mid_c = tl.minimum(mid, nbins)
+            edge = tl.load(row + mid_c, mask=valid, other=float("inf")).to(ACC_DTYPE)
+            take = (edge <= coord) & (mid <= nbins)
+            lo = tl.where(take, mid, lo)
+            hi = tl.where(take, hi, mid - 1)
+        bin_idx = lo
+
+        # The rightmost bin is inclusive of its right edge, so a point exactly on
+        # `right` searches to index nbins and must fold back into the last bin.
+        bin_idx = tl.where(bin_idx >= nbins, nbins - 1, bin_idx)
```
Comment (Caeruleann) [inline src/flag_gems/ops/_histogramdd_from_bin_cts.py:115]: For a zero-width explicit range like [1.0, 1.0], ATen puts in-range points into bin nbins//2, but the binary search folds them into the last bin (nbins-1). Please special-case width == 0 to match ATen and add a parity test.

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 9 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py]: 已修（commit dffdebc7）。host 侧的 `torch.isfinite` / `nonzero` 去掉了，有限性判断现在整个在设备侧完成。

`histogramdd_col_range_kernel` 增加一个单元素的 `nonfinite_ptr` 输出：

- 每个 program 在自己的列 tile 内用一个独立的布尔累加器 `nonfinite_seen` 跨 row tile 做 OR（NaN 或 ±Inf 都算），然后 `tl.min` 取出本 tile 内最小的越界维度；
- 再 `tl.atomic_min` 归约到 `nonfinite_ptr`，即全局最小越界维度（`(~finite).nonzero()[0]` 的语义）；
- host 侧初始化成哨兵 `N`（大于任何合法维度），kernel 只做「取小」，所以全有限时哨兵原样保留。`_check_finite_auto_range` 现在只读这**一个标量**：

```python
dim = int(nonfinite_dim.item())
if dim >= N:
    return
```

顺带说明：`nonfinite_seen` 与原有的 `nan_seen` 是分开的，因为两者语义不同——`nan_seen` 决定 range 是否写成 NaN（要跟 `aminmax` 一致），而 ±Inf 会静默被 min/max 累加器吞掉、并不产生 NaN，但同样让 range 非有限、同样要被拒绝。

验证：NaN/±Inf × 3 个列位置 × 3 个 row 位置（含跨 BLOCK_M=128 的 tile 边界）共 27/27 与 native 消息逐字一致；正常有限数据不受影响。

新增 `test_histogramdd_from_bin_cts_auto_range_reports_lowest_dimension`：一次污染两个及以上列，断言报出的是**最低**那个维度而不是最后一个或 tile 内局部的一个——否则回归了也照样 raise，只是消息里的维度是错的。

File: src/flag_gems/ops/_histogramdd_from_bin_cts.py (hunk 10 above)
Comment (Yukun-Cui) [reply src/flag_gems/ops/_histogramdd_from_bin_cts.py:116]: 已修（commit dffdebc7）。

先纠正一点：问题不在「折叠到最后一个 bin」，而在**根本不该走到搜索逻辑**。

读了 ATen 的 `select_outer_bin_edges`（`aten/src/ATen/native/Histogram.cpp`），显式 range 也要过一遍逐维校验，其中包含：

```cpp
// Expand empty range to match numpy behavior and avoid division by 0 in normalization
if (leftmost_edge == rightmost_edge) {
  leftmost_edges[dim] -= 0.5;
  rightmost_edges[dim] += 0.5;
}
```

也就是说 `range=[1.0, 1.0]` 在构造 edges **之前**就被展宽成 `[0.5, 1.5]`，edges 实际是 `linspace(0.5, 1.5, nbins+1)`——与退化 auto-range（min==max）走的是同一条路径。实测确认：

```
torch._histogramdd_bin_edges(pts, [nbins], range=[1.0,1.0])[0] == torch.linspace(0.5,1.5,nbins+1)   # nbins=1..10 全部 True
```

这也解释了 `nbins//2` 的来源：不是特判，而是 `1.0` 正好落在展宽后区间的中间 bin 的边界上，搜索自然选中它。

所以修法是在 `_resolve_range` 里，逐维校验之后、构造 edges 之前，对零宽维度做同样的 `±0.5` 展宽。实测（1-D，点取 1.0/1.0/5.0）：

```
nbins=1   gems=[2.0]                                    aten=[2.0]
nbins=2   gems=[0.0, 2.0]                               aten=[0.0, 2.0]
nbins=3   gems=[0.0, 2.0, 0.0]                          aten=[0.0, 2.0, 0.0]
nbins=4   gems=[0.0, 0.0, 2.0, 0.0]                     aten=[0.0, 0.0, 2.0, 0.0]
nbins=5   gems=[0.0, 0.0, 2.0, 0.0, 0.0]                aten=[0.0, 0.0, 2.0, 0.0, 0.0]
nbins=7   gems=[0.0, 0.0, 0.0, 2.0, 0.0, 0.0, 0.0]      aten=[0.0, 0.0, 0.0, 2.0, 0.0, 0.0, 0.0]
nbins=10  gems=[0,0,0,0,0,2.0,0,0,0,0]                  aten=[0,0,0,0,0,2.0,0,0,0,0]
```

5.0 在展宽范围之外，保持不计入。

顺带发现并修了一个同源的洞：显式 range 传非有限值（如 `[inf, inf]`、`[nan, 1.0]`）原本静默返回垃圾直方图，ATen 同样会拒绝。而且 ATen 的**校验顺序是「有限性」先于「min <= max」**，实测 `[inf, -inf]` 报的是 "is not finite" 而不是 "min should not exceed max"，所以两处检查按这个顺序排。8 组 range 与 ATen 逐字一致。

新增测试：
- `test_histogramdd_from_bin_cts_zero_width_explicit_range`（nbins 1/2/3/4/5/7/10，断言中间 bin 精确计数 + 越界点不计入）
- `test_histogramdd_from_bin_cts_zero_width_explicit_range_2d`（一个维度零宽、另一个正常）
- `test_histogramdd_from_bin_cts_zero_width_edges_match_linspace`（展宽后的边界本身，含边界两侧的探针）
- `test_histogramdd_from_bin_cts_rejects_non_finite_explicit_range`（5 组，与 native 消息逐字比对）
- `test_histogramdd_from_bin_cts_finite_check_precedes_min_max`（钉住检查顺序）

`tests/test_histogramdd_from_bin_cts.py`: 116 passed（`--ref=cpu --quick` 80 passed）。

**Review by Caeruleann — APPROVED** [review]
